include/nbl/video/IGPUCommandPool.h

```diff
@@ -31,13 +66,219 @@ class IGPUCommandPool : public core::IReferenceCounted, public IBackendObject
         // Vulkan: const VkCommandPool*
         virtual const void* getNativeHandle() const = 0;
 
+        core::CMemoryPool<core::GeneralpurposeAddressAllocator<uint32_t>, core::default_aligned_allocator, false, uint32_t> m_commandSegmentPool;
+
     protected:
         virtual ~IGPUCommandPool() = default;
 
         core::bitflag<E_CREATE_FLAGS> m_flags;
         uint32_t m_familyIx;
 };
 
+class alignas(COMMAND_ALIGNMENT) IGPUCommandPool::ICommand
+{
+public:
+    virtual ~ICommand() {}
+
+    // static void* operator new(std::size_t size) = delete;
+    static void* operator new[](std::size_t size) = delete;
+    // static void* operator new(std::size_t size, std::align_val_t al) = delete;
+    static void* operator new[](std::size_t size, std::align_val_t al) = delete;
+
+    // static void operator delete  (void* ptr) = delete;
+    static void operator delete[](void* ptr) = delete;
+    static void operator delete  (void* ptr, std::align_val_t al) = delete;
+    static void operator delete[](void* ptr, std::align_val_t al) = delete;
+    static void operator delete  (void* ptr, std::size_t sz) = delete;
+    static void operator delete[](void* ptr, std::size_t sz) = delete;
+    static void operator delete  (void* ptr, std::size_t sz, std::align_val_t al) = delete;
+    static void operator delete[](void* ptr, std::size_t sz, std::align_val_t al) = delete;
+
+    uint32_t m_size;
+protected:
+    ICommand(uint32_t size) : m_size(size) {}
+};
+
+class alignas(COMMAND_SEGMENT_ALIGNMENT) IGPUCommandPool::CommandSegment
+{
+public:
+    struct Iterator;
+
+    struct params_t
+    {
+        core::LinearAddressAllocator<uint32_t> m_commandAllocator;
+        CommandSegment* m_next;
+    } params;
+
+    static inline constexpr uint32_t STORAGE_SIZE = COMMAND_SEGMENT_SIZE - sizeof(params_t);
+    alignas(ICommand) uint8_t m_data[STORAGE_SIZE];
+
+    CommandSegment()
+    {
+        params.m_commandAllocator = core::LinearAddressAllocator<uint32_t>(nullptr, 0u, 0u, COMMAND_SEGMENT_ALIGNMENT, COMMAND_SEGMENT_SIZE);
```

Comment:
for readability `alignof(m_data)` would be better than `COMMAND_SEGMENT_ALIGNMENT`

Reply:
They could technically be different values, so then `static_assert(alignof(m_data) == COMMAND_SEGMENT_ALIGNMENT)`?

Reply:
yes that can be asserted

Reply:
`alignof(m_data)` is 1.
I think you mean `alignof(ICommand)`?

Reply:
> They could technically be different values, so then `static_assert(alignof(m_data) == COMMAND_SEGMENT_ALIGNMENT)`?

Also, I don't think we need to assert this because the `maxAllocatableAlignment` for `LinearAddressAllocator` (specifically used to allocate `ICommand`s) should be `alignof(ICommand)`?

Reply:
there's nothing wrong with peppering asserts everywhere

Reply:
Resolved.